ARCs/arc-0058.md

```diff
@@ -0,0 +1,520 @@
+
+This requires wallets to be able to tell apps that the connected account is an Abstracted Account.
+
+## Rationale
+
+### App vs Logic Sig
+There have similar propsoals for reducing end-user friction, such as [ARC-47](./arc-0047.md) which enables safer usage of delegated logic signatures. The major downside of logic signatures is that they are not usable by smart contracts. This severely limits composability and potential use cases.
+
+### Plugins
+Rather than constantly updating the approval program of the abstracted account application to add functionality, it is safer and easier to simply add additional apps that enable the desired functionality. This also gives the end-user more control over what various dApps can do with their account at any time.
+
+### Plugin Permisions
+
+A common use case for plugins will be end-users allowing specific apps to perform actions on their account. As such, implementing this in the Abstracted Account App allows for wallets and other ecosystem tools to easy display permissions to end users. The end time is also useful to ensure a user only enables a plugin for the time that it would be useful for them. The concept of plugins is similar to approvals on EVM chains and there have been cases where old approvals became an attack vector.
+
+## Backwards Compatibility
+Existing Algorand accounts can transition to an abstracted account by creating a new abstracted account application and setting the address to their current address. This requires them to create a new account to act as the admin.
+
+End-users can use an abstracted account with any dApp provided they rekey the account to an externally owned account.
+
+## Test Cases
+
+TODO: Some functional tests are in this repo https://github.com/joe-p/account_abstraction.git
+
+## Reference Implementation
+
+### Abstracted App
+
+```ts
+import { Contract } from '@algorandfoundation/tealscript';
+
+type PluginsKey = { application: Application; allowedCaller: Address };
+
+export class AbstractedAccount extends Contract {
+  /** Target AVM 10 */
+  programVersion = 10;
+
+  /** The admin of the abstracted account */
+  admin = GlobalStateKey<Address>();
+
+  /**
+   * The apps and addresses that are authorized to send itxns from the abstracted account,
+   * The key is the appID + address, the value (referred to as `end`)
+   * is the timestamp when the permission expires for the address to call the app for your account.
+   */
+
+  plugins = BoxMap<PluginsKey, uint64>({ prefix: 'p' });
+
+  /**
+   * Plugins that have been given a name for discoverability
+   */
+  namedPlugins = BoxMap<bytes, PluginsKey>({ prefix: 'n' });
+
+  /** The address this app controls */
+  controlledAddress = GlobalStateKey<Address>();
+
+  /**
+   * Ensure that by the end of the group the abstracted account has control of its address
+   */
+  private verifyRekeyToAbstractedAccount(): void {
+    const lastTxn = this.txnGroup[this.txnGroup.length - 1];
```

Comment:
For better composability, could we make this more dynamic and allow the rekey to happen before the last transaction in a group?

As an example, one could do it like Tinyman V2 does it in its flash loans: they have two methods`flash_loan` and `verify_flash_loan` that reference each other and check that the counterpart exists.

Reply:
Hey! If you look just below this comment you'll see the implementation does allow for rekeys earlier in the group and if so it just asserts that its been rekeyed back properly

Reply:
Yeah I would imagine `arc58_verifyAuthAddr` is the equivalent to  `verify_flash_loan`. The last txn must either be an explicit rekey or a call to `arc58_verifyAuthAddr`.

Reply:
Thanks for the clarification. The difference is that `verify_flash_loan` doesn't have to be the last transaction in the group, while here either an explicit rekeying or `arc58_verifyAuthAddr` have to be the last transaction in the group (if I understand it correctly). My point is that for better composability it would be nice if one could take ARC-58 transactions and freely add more transactions in front or at the end of a group.

Reply:
Awesome! One question: do you think running out of opcode budget could be a problem because of the loop in case the transaction group is big? A small optimization could be to start the for loop at one past the current group index.

Reply:
Ah yeah good catch on the starting index. I was originally thinking that additional txns in the group would be app calls but I suppose that's not always the case. I will add a worst-case scenario to my tests and see if the budget is exceeded.